Move AXObjectCache from Document to Page

achristensen07 opened the pull request:
#### 49d79a2453dec558b914e4e8ea8158b3adeed431
<pre>
Move AXObjectCache from Document to Page
<a href="https://bugs.webkit.org/show_bug.cgi?id=286199">https://bugs.webkit.org/show_bug.cgi?id=286199</a>

Reviewed by Tyler Wilcock.

There is only one per main Document, which means there is
one per Page anyways.  This is preparation for better management
of AXObjectCache in site isolated processes where the main
Frame is a RemoteFrame.

There is one test that covers this with site isolation on
in a case where a Page has no main frame Document,
http/tests/site-isolation/accessibility/hit-test-resolving-remote-frame.html.
To get this test to continue working, I needed to make
WKAccessibilityWebPageObjectBase get the AXObjectCache directly
from the Page instead of the main frame&apos;s Document, and I needed
to make accessibilityRootObjectWrapper take an optional LocalFrame
and use the LocalFrame to call AXObjectCache::rootObjectForFrame
instead of AXObjectCache::rootObject.

* Source/WebCore/dom/Document.cpp:
(WebCore::Document::createRenderTree):
(WebCore::Document::clearAXObjectCache):
(WebCore::Document::existingAXObjectCacheSlow const):
(WebCore::Document::axObjectCache const):
* Source/WebCore/dom/Document.h:
* Source/WebCore/page/Page.cpp:
(WebCore::Page::clearAXObjectCache):
(WebCore::Page::axObjectCache):
* Source/WebCore/page/Page.h:
(WebCore::Page::existingAXObjectCache):

Canonical link: <a href="https://commits.webkit.org/289258@main">https://commits.webkit.org/289258@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/15da4d7a0442cd4b3ae0e322bba23e29debadce2

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/85913 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/131/builds/5534 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/55/builds/40281 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/90923 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/36824 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/87958 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/130/builds/5773 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/13538 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/66678 "Passed tests") | [✅ 🧪 win-tests](https://ews-build.webkit.org/#/builders/60/builds/24472 "Passed tests") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/88916 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/132/builds/4418 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/77926 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/46968 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/133/builds/4273 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/64/builds/32212 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/35904 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/74905 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/63/builds/33073 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/92693 "Built successfully") | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/128/builds/13162 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/122/builds/9654 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/75455 "Passed tests") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/121/builds/13376 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/73777 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/74610 "Passed tests") | 
| | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/126/builds/18797 "Passed tests") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/119/builds/17236 "Passed tests") | [  ~~🛠 playstation~~](https://ews-build.webkit.org/#/builders/134/builds/6044 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| [✅ 🛠 🧪 unsafe-merge](https://ews-build.webkit.org/#/builders/22/builds/13400 "Built successfully and passed tests") | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/127/builds/13195 "Built successfully") | [✅ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/18543 "Built successfully") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/125/builds/12968 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/129/builds/16401 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/124/builds/14755 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

hoffmanjoshua commented on Source/WebCore/page/Page.cpp:
```diff
@@ -4068,6 +4068,24 @@ void Page::appearanceDidChange()
+{
+    if (!m_axObjectCache) {
+        RefPtr localMainFrame = dynamicDowncast<LocalFrame>(m_mainFrame.get());
+        RefPtr mainFrameDocument = localMainFrame ? localMainFrame->document() : nullptr;
```
If the main frame is a `RemoteFrame`, we won't have a `mainFrameDocument`, correct? Curious what the implications will be then, since the cache will not have a document to initialize with.

achristensen07 replied:
Correct.  We will need to go to all such cases and replace the use of AXObjectCache::document with specifying which document to use.

hoffmanjoshua commented on Source/WebKit/WebProcess/WebPage/mac/WKAccessibilityWebPageObjectBase.mm:
```diff
@@ -106,14 +98,16 @@ - (id)accessibilityRootObjectWrapper
             return protectedSelf.get().accessibilityPluginObject;
 
         if (auto cache = protectedSelf.get().axObjectCache) {
+            if (auto* root = cache->rootObjectForFrame(frame.get()))
```
Is the goal here to have more requests specify a frame, rather that just get the root object for the entire cache?

achristensen07 replied:
Yes, that will be necessary when enabling site isolation and having pages with no main frame document

achristensen07 committed:
Move AXObjectCache from Document to Page
https://bugs.webkit.org/show_bug.cgi?id=286199

Reviewed by Tyler Wilcock.

There is only one per main Document, which means there is
one per Page anyways.  This is preparation for better management
of AXObjectCache in site isolated processes where the main
Frame is a RemoteFrame.

There is one test that covers this with site isolation on
in a case where a Page has no main frame Document,
http/tests/site-isolation/accessibility/hit-test-resolving-remote-frame.html.
To get this test to continue working, I needed to make
WKAccessibilityWebPageObjectBase get the AXObjectCache directly
from the Page instead of the main frame's Document, and I needed
to make accessibilityRootObjectWrapper take an optional LocalFrame
and use the LocalFrame to call AXObjectCache::rootObjectForFrame
instead of AXObjectCache::rootObject.

* Source/WebCore/dom/Document.cpp:
(WebCore::Document::createRenderTree):
(WebCore::Document::clearAXObjectCache):
(WebCore::Document::existingAXObjectCacheSlow const):
(WebCore::Document::axObjectCache const):
* Source/WebCore/dom/Document.h:
* Source/WebCore/page/Page.cpp:
(WebCore::Page::clearAXObjectCache):
(WebCore::Page::axObjectCache):
* Source/WebCore/page/Page.h:
(WebCore::Page::existingAXObjectCache):

Canonical link: https://commits.webkit.org/289258@main
